// algob/internal/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BuilderContext = void 0;
const extenders_1 = require("./core/config/extenders");
const errors_1 = require("./core/errors");
const errors_list_1 = require("./core/errors-list");
const dsl_1 = require("./core/tasks/dsl");
class BuilderContext {
    constructor() {
        this.tasksDSL = new dsl_1.TasksDSL();
        this.extendersManager = new extenders_1.ExtenderManager();
        this.loadedPlugins = [];
        this.configExtenders = [];
    }
    static isCreated() {
        const globalWithBuilderContext = global;
        return globalWithBuilderContext.__builderContext !== undefined;
    }
    static createBuilderContext() {
        if (this.isCreated()) {
            throw new errors_1.BuilderError(errors_list_1.ERRORS.GENERAL.CONTEXT_ALREADY_CREATED);
        }
        const globalWithBuilderContext = global;
        const ctx = new BuilderContext();
        globalWithBuilderContext.__builderContext = ctx;
        return ctx;
    }
    static getBuilderContext() {
        const globalWithBuilderContext = global;
        const ctx = globalWithBuilderContext.__builderContext;
        if (ctx === undefined) {
            throw new errors_1.BuilderError(errors_list_1.ERRORS.GENERAL.CONTEXT_NOT_CREATED);
        }
        return ctx;
    }
    static deleteBuilderContext() {
        global.__builderContext = undefined; // eslint-disable-line @typescript-eslint/no-explicit-any
    }
    setAlgobRuntimeEnv(env) {
        if (this.environment !== undefined) {
            throw new errors_1.BuilderError(errors_list_1.ERRORS.GENERAL.CONTEXT_BRE_ALREADY_DEFINED);
        }
        this.environment = env;
    }
    getAlgobRuntimeEnv() {
        if (this.environment === undefined) {
            throw new errors_1.BuilderError(errors_list_1.ERRORS.GENERAL.CONTEXT_BRE_NOT_DEFINED);
        }
        return this.environment;
    }
    setPluginAsLoaded(pluginName) {
        this.loadedPlugins.push(pluginName);
    }
}
exports.BuilderContext = BuilderContext;
