// algob/internal/tx-log-writer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TxWriterImpl = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const yaml_1 = __importDefault(require("yaml"));
const project_structure_1 = require("./core/project-structure");
class TxWriterImpl {
    constructor(scriptName) {
        this.timestamp = +new Date();
        this.scriptName = scriptName;
    }
    setScriptName(scriptName) {
        this.scriptName = scriptName;
    }
    ensureDirectoryExistence(filePath) {
        var dirname = path_1.default.dirname(filePath);
        if (fs_1.default.existsSync(dirname)) {
            return;
        }
        this.ensureDirectoryExistence(dirname);
        fs_1.default.mkdirSync(dirname);
    }
    push(msg, obj) {
        this.timestamp = +new Date();
        const filePath = path_1.default.join(project_structure_1.ARTIFACTS_DIR, this.scriptName) + '.' +
            (this.timestamp).toString() + '.log';
        this.ensureDirectoryExistence(filePath);
        var map = new Map();
        map.set(msg, obj);
        yaml_1.default.defaultOptions.indent = 4;
        var file = yaml_1.default.stringify(map);
        fs_1.default.appendFileSync(filePath, file);
    }
}
exports.TxWriterImpl = TxWriterImpl;
